File: src/Core/Templates.fs
namespace ARCtrl

open ARCtrl
open ARCtrl.Helper
open Fable.Core

/// This module contains unchecked helper functions for templates
[<RequireQualifiedAccess>]
module TemplatesAux =

    let getComparer (matchAll: bool option) =
        let matchAll = defaultArg matchAll false
        let comparer = if matchAll then (&&) else (||)
        comparer

    let filterOnTags (tagGetter: Template -> OntologyAnnotation ResizeArray) (queryTags: OntologyAnnotation ResizeArray) (comparer: bool -> bool -> bool) (templates: Template ResizeArray) =
        templates 
        |> ResizeArray.filter(fun t ->
            let templateTags = tagGetter t
            let mutable isValid = None
            for qt in queryTags do
                let contains = templateTags |> Seq.contains qt
                match isValid, contains with
                | None, any -> isValid <- Some any
                | Some maybe, any -> isValid <- Some (comparer maybe any)
            Option.defaultValue false isValid
        )

[<AttachMembers>]
type Templates =

    static member getDistinctTags (templates: Template ResizeArray) =
        templates |> ResizeArray.collect (fun t -> t.Tags) |> ResizeArray.distinct

    /// <summary>
    /// Returns all **distinct** `template.Tags` and `template.EndpointRepositories`
    /// </summary>
    /// <param name="templates"></param>
    static member getDistinctEndpointRepositories (templates: Template ResizeArray) =
        templates |> ResizeArray.collect (fun t -> t.EndpointRepositories) |> ResizeArray.distinct

    /// <summary>
    /// Returns all **distinct** `template.Tags` and `template.EndpointRepositories`
    /// </summary>
    /// <param name="templates"></param>
    static member getDistinctOntologyAnnotations (templates: Template []) =
        let oas = ResizeArray()
        for t in templates do
            oas.AddRange(t.Tags)
            oas.AddRange(t.EndpointRepositories)
        oas
        |> Array.ofSeq
        |> Array.distinct

    /// <summary>
    /// Filter templates by `template.Tags`.
    /// </summary>
    /// <param name="queryTags">The ontology annotation to filter by.</param>
    /// <param name="matchAll">Default: false. If true all `queryTags` must be contained in template, if false only 1 tags must be contained in template.</param>
    static member filterByTags(queryTags: OntologyAnnotation ResizeArray, ?matchAll: bool) = 
        fun (templates: Template ResizeArray) ->
            let comparer = TemplatesAux.getComparer matchAll
            TemplatesAux.filterOnTags (fun t -> t.Tags) queryTags comparer templates

    /// <summary>
    /// Filter templates by `template.EndpointRepositories`.
    /// </summary>
    /// <param name="queryTags">The ontology annotation to filter by.</param>
    /// <param name="matchAll">Default: false. If true all `queryTags` must be contained in template, if false only 1 tags must be contained in template.</param>
    static member filterByEndpointRepositories(queryTags: OntologyAnnotation ResizeArray, ?matchAll: bool) = 
        fun (templates: Template ResizeArray) ->
            let comparer = TemplatesAux.getComparer matchAll
            TemplatesAux.filterOnTags (fun t -> t.EndpointRepositories) queryTags comparer templates

    /// <summary>
    /// Filters templates by template.Tags and template.EndpointRepositories
    /// </summary>
    /// <param name="queryTags">The ontology annotation to filter by.</param>
    /// <param name="matchAll">Default: false. If true all `queryTags` must be contained in template, if false only 1 tags must be contained in template.</param>
    static member filterByOntologyAnnotation(queryTags: OntologyAnnotation ResizeArray, ?matchAll: bool) = 
        fun (templates: Template ResizeArray) ->
            let comparer = TemplatesAux.getComparer matchAll
            TemplatesAux.filterOnTags (fun t -> ResizeArray.append t.Tags t.EndpointRepositories) queryTags comparer templates

    /// <summary>
    /// Filters templates by template.Organisation = `Organisation.DataPLANT`/`"DataPLANT"`.
    /// </summary>
    /// <param name="templates"></param>
    static member filterByDataPLANT (templates: Template ResizeArray) =
        templates
        |> ResizeArray.filter (fun t -> t.Organisation.IsOfficial())
        

File: js/Core/Templates.js
import { defaultArg } from "../fable_modules/fable-library-js.4.22.0/Option.js";
import { ResizeArray_append, ResizeArray_collect, ResizeArray_distinct, ResizeArray_filter } from "./Helper/Collections.js";
import { uncurry2, safeHash, equals, disposeSafe, getEnumerator } from "../fable_modules/fable-library-js.4.22.0/Util.js";
import { contains as contains_1 } from "../fable_modules/fable-library-js.4.22.0/Seq.js";
import { addRangeInPlace, item } from "../fable_modules/fable-library-js.4.22.0/Array.js";
import { Array_distinct } from "../fable_modules/fable-library-js.4.22.0/Seq2.js";
import { class_type } from "../fable_modules/fable-library-js.4.22.0/Reflection.js";

export function TemplatesAux_getComparer(matchAll) {
    if (defaultArg(matchAll, false)) {
        return (e) => ((e_1) => (e && e_1));
    }
    else {
        return (e_2) => ((e_3) => (e_2 || e_3));
    }
}

export function TemplatesAux_filterOnTags(tagGetter, queryTags, comparer, templates) {
    return ResizeArray_filter((t) => {
        const templateTags = tagGetter(t);
        let isValid = undefined;
        let enumerator = getEnumerator(queryTags);
        try {
            while (enumerator["System.Collections.IEnumerator.MoveNext"]()) {
                const contains = contains_1(enumerator["System.Collections.Generic.IEnumerator`1.get_Current"](), templateTags, {
                    Equals: equals,
                    GetHashCode: safeHash,
                });
                const isValid_1 = isValid;
                if (isValid_1 != null) {
                    const maybe = isValid_1;
                    isValid = comparer(maybe, contains);
                }
                else {
                    isValid = contains;
                }
            }
        }
        finally {
            disposeSafe(enumerator);
        }
        return defaultArg(isValid, false);
    }, templates);
}

export class Templates {
    constructor() {
    }
    static getDistinctTags(templates) {
        return ResizeArray_distinct(ResizeArray_collect((t) => t.Tags, templates));
    }
    static getDistinctEndpointRepositories(templates) {
        return ResizeArray_distinct(ResizeArray_collect((t) => t.EndpointRepositories, templates));
    }
    static getDistinctOntologyAnnotations(templates) {
        const oas = [];
        for (let idx = 0; idx <= (templates.length - 1); idx++) {
            const t = item(idx, templates);
            addRangeInPlace(t.Tags, oas);
            addRangeInPlace(t.EndpointRepositories, oas);
        }
        return Array_distinct(Array.from(oas), {
            Equals: equals,
            GetHashCode: safeHash,
        });
    }
    static filterByTags(queryTags, matchAll) {
        return (templates) => TemplatesAux_filterOnTags((t) => t.Tags, queryTags, uncurry2(TemplatesAux_getComparer(matchAll)), templates);
    }
    static filterByEndpointRepositories(queryTags, matchAll) {
        return (templates) => TemplatesAux_filterOnTags((t) => t.EndpointRepositories, queryTags, uncurry2(TemplatesAux_getComparer(matchAll)), templates);
    }
    static filterByOntologyAnnotation(queryTags, matchAll) {
        return (templates) => TemplatesAux_filterOnTags((t) => ResizeArray_append(t.Tags, t.EndpointRepositories), queryTags, uncurry2(TemplatesAux_getComparer(matchAll)), templates);
    }
    static filterByDataPLANT(templates) {
        return ResizeArray_filter((t) => t.Organisation.IsOfficial(), templates);
    }
}

export function Templates_$reflection() {
    return class_type("ARCtrl.Templates", undefined, Templates);
}

